src/plugins/hidsim/gui_monitor.cpp

```diff
@@ -0,0 +1,443 @@
+ * license file for more details (it's in a COPYING file included with     *
+ * DRAKVUF, and also available from                                        *
+ * https://github.com/tklengyel/drakvuf/COPYING)                           *
+ *                                                                         *
+ * This file was created by Jan Gruber.                                    *
+ * It is distributed as part of DRAKVUF under the same license             *
+ ***************************************************************************/
+
+#include <unistd.h>
+#include <sys/time.h>
+
+#include <libdrakvuf/libdrakvuf.h>  /* eprint_current_time */
+#include <json-c/json_object.h>     /* json_object_put,... */
+
+/* Plugin code */
+#include "../plugins.h"
+#include "../private.h"             /* PRINT_DEBUG */
+
+/* GUI reconstruction specific code */
+#include "gui/vmi_win_gui_offsets.h"
+#include "gui/vmi_win_gui_parser.h"
+#include "gui/vmi_win_gui_utils.h"
+
+#include "gui_monitor.h"
+
+#define TRAP_FUNC "NtUserShowWindow"
+
+/* Signifies, whether a GUI update has occured */
+static volatile sig_atomic_t has_gui_update = false;
+
+/* Timestamp of last drawing action */
+static struct timeval last;
+
+/* Synchronizes access to struct timeval last */
+static std::mutex lock;
+
+/* Trap on NtUserShowWindow */
+static drakvuf_trap_t gui_trap =
+{
+    .breakpoint.lookup_type = LOOKUP_DTB,
+    .breakpoint.addr_type = ADDR_VA,
+    .type = BREAKPOINT,
+    .data = NULL,
+};
+
+/* Callback triggered on NtUserShowWindow(...) */
+static event_response_t on_draw(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+    has_gui_update = true;
+
+    lock.lock();
+    /* Keeps track of the last event */
+    gettimeofday(&last, NULL);
+    lock.unlock();
+
+    return VMI_EVENT_RESPONSE_NONE;
+}
+/*
+ * Registers a trap for the system call for NtUserShowWindow of win32k.sys by
+ * utilizing the following routine:
+ *
+ *     1.  Find RVA of the function of interest (here NtUserShowWindow) from
+ *         win32k-IST-JSON
+ *     2.  Find RVA of the GUI SSDT called W32pServiceTable from
+ *         win32k-IST-JSON
+ *     3.  Find VA of KeServiceDescriptorTableShadow-symbol
+ *     4.  Retrieve DTB of an arbitrary interactive process by reading CR3
+ *     5.  Use this DTB to resolve the pointer to the GUI SSDT
+ *     6.  Find the address to trap by subtracting the RVA of W32pServiceTable
+ *         from the VA of the GUI SSDT and then adding the function offset
+ *
+ * For a thorough description of the inner workings of the SSDT and
+ * the W32pServiceTable (aka GUI SSDT), see ERNWs blogpost at
+ * https://insinuator.net/2015/12/\
+ * investigating-memory-analysis-tools-ssdt-hooking-via-pointer-replacement/
+ */
+static bool register_trap( drakvuf_t drakvuf, json_object* profile_json,
+    const char* function_name, drakvuf_trap_t* trap,
+    event_response_t(*hook_cb)( drakvuf_t drakvuf, drakvuf_trap_info_t* info))
+{
```

Comment:
1) I'm 99.9% sure `register_trap` will shadow some already existing function name and cause confusion, please rename it.
2) I'm 99.9% sure `event_response_t` type is already defined somewhere so you don't need to redefine it.

Reply:
> I'm 99.9% sure `register_trap` will shadow some already existing function name and cause confusion, please rename it.
 
Changed.
 
 >  I'm 99.9% sure `event_response_t` type is already defined somewhere so you don't need to redefine it.
 
`event_response_t` is defined in `libvmi.h`. However, since I'm passing a function pointer, I don't see another way, than defining the function signature of the expected function pointer, which should be passed as an argument to this function.  
Please give me some direction...

Reply:
We usually use: https://github.com/tklengyel/drakvuf/blob/03a21caedf63a086f2f64683f56efecf5cbc1226/src/plugins/plugins_ex.h#L413 typedef, but now I see it's in pluginex not in plugin. Imo it could easily be moved to superclass.

Reply:
> now I see it's in pluginex not in plugin. Imo it could easily be moved to superclass.

Since this is `gui_monitor.cpp` and not in the actual plugin-class: Do you think a change would have a real benefit or could you live with the current implementation?

Reply:
Yep, fine with me.